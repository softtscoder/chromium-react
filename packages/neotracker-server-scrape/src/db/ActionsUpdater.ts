import { addressToScriptHash, RawAction } from '@neo-one/client-full';
import { createChild, serverLogger } from '@neotracker/logger';
import { Action as ActionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { strip0x } from '../utils';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface ActionsSaveSingle {
  readonly action: RawAction;
  readonly transactionID: string;
  readonly transactionHash: string;
}
export interface ActionsSave {
  readonly actions: ReadonlyArray<ActionsSaveSingle>;
}
export interface ActionsRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class ActionsUpdater extends SameContextDBUpdater<ActionsSave, ActionsRevert> {
  public async save(context: Context, { actions }: ActionsSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_actions' });
    await Promise.all(
      _.chunk(actions, context.chunkSize).map(async (chunk) =>
        ActionModel.insertAll(
          context.db,
          context.makeQueryContext(),
          chunk.map(({ action, transactionID, transactionHash }) => ({
            id: action.globalIndex.toString(),
            type: action.type,
            block_id: action.blockIndex,
            transaction_id: transactionID,
            transaction_hash: transactionHash,
            transaction_index: action.transactionIndex,
            index: action.index,
            script_hash: strip0x(addressToScriptHash(action.address)),
            // tslint:disable no-any no-null-keyword
            message: (action as any).message === undefined ? null : encodeURIComponent((action as any).message),
            args_raw: (action as any).args === undefined ? null : JSON.stringify((action as any).args),
            // tslint:disable-enable no-any no-null-keyword
          })),
        ),
      ),
    );
  }

  public async revert(context: Context, { transactionIDs }: ActionsRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_actions' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map(async (chunk) =>
        ActionModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('transaction_id', chunk)
          .delete(),
      ),
    );
  }
}
