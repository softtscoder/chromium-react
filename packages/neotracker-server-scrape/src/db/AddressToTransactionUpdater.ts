import { createChild, serverLogger } from '@neotracker/logger';
import { AddressToTransaction as AddressToTransactionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressToTransactionSaveSingle {
  readonly addressIDs: ReadonlyArray<string>;
  readonly transactionID: string;
}
export interface AddressToTransactionSave {
  readonly transactions: ReadonlyArray<AddressToTransactionSaveSingle>;
}
export interface AddressToTransactionRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressToTransactionUpdater extends SameContextDBUpdater<
  AddressToTransactionSave,
  AddressToTransactionRevert
> {
  public async save(context: Context, { transactions }: AddressToTransactionSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_address_to_transaction' });
    const data = _.flatMap(transactions, ({ addressIDs, transactionID }) =>
      [...new Set(addressIDs)].map((addressID) => ({
        id1: addressID,
        id2: transactionID,
      })),
    );
    await Promise.all(
      _.chunk(data, context.chunkSize).map(async (chunk) => {
        await AddressToTransactionModel.insertAll(context.db, context.makeQueryContext(), chunk);
      }),
    );
  }

  public async revert(context: Context, { transactionIDs }: AddressToTransactionRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_address_to_transaction' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map((chunk) =>
        AddressToTransactionModel.query(context.db)
          .context(context.makeQueryContext())
          .delete()
          .whereIn('id2', chunk),
      ),
    );
  }
}
