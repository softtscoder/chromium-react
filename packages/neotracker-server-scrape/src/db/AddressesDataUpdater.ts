import { createChild, serverLogger } from '@neotracker/logger';
import { Address as AddressModel, isPostgres } from '@neotracker/server-db';
import { raw } from 'objection';
import { Addresses, Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressesDataSave {
  readonly addresses: Addresses;
  readonly blockIndex: number;
  readonly blockTime: number;
}
export interface AddressesDataRevert {
  readonly addresses: Addresses;
  readonly transactionIDs: ReadonlyArray<string>;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressesDataUpdater extends SameContextDBUpdater<AddressesDataSave, AddressesDataRevert> {
  public async save(context: Context, { addresses, blockIndex, blockTime }: AddressesDataSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(
        async ([address, { transactionCount, transferCount, transactionID, transactionHash }]) => {
          await AddressModel.query(context.db)
            .context(context.makeQueryContext())
            .where('id', address)
            .where('aggregate_block_id', '<', blockIndex)
            .patch({
              // tslint:disable no-any
              transaction_count: raw(`transaction_count + ${transactionCount}`) as any,
              transfer_count: raw(`transfer_count + ${transferCount}`) as any,
              // tslint:enable no-any
              last_transaction_id: transactionID,
              last_transaction_hash: transactionHash,
              last_transaction_time: blockTime,
              aggregate_block_id: blockIndex,
            });
        },
      ),
    );
  }

  public async revert(context: Context, { addresses, transactionIDs, blockIndex }: AddressesDataRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(async ([address, { transactionCount, transferCount }]) => {
        await AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', address)
          .where('aggregate_block_id', '>=', blockIndex)
          .patch({
            // tslint:disable no-any
            transaction_count: raw(`transaction_count - ${transactionCount}`) as any,
            transfer_count: raw(`transfer_count - ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex - 1,
          });
      }),
    );

    const addressIDsSet = Object.keys(addresses);
    if (addressIDsSet.length > 0) {
      if (isPostgres(context.db)) {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.id AS address_id, b.transaction_id, b.hash, b.block_time
                FROM address a
                LEFT OUTER JOIN (
                  SELECT address_id, transaction_id, hash, block_time
                  FROM (
                    SELECT id1 AS address_id, MAX(id2) AS transaction_id
                    FROM address_to_transaction
                    WHERE
                      id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                      id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                    GROUP BY address_to_transaction.id1
                  ) a
                  JOIN transaction b ON
                    a.transaction_id = b.id
                ) b ON
                  a.id = b.address_id
                WHERE
                  a.id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
              )
              UPDATE address SET
                last_transaction_id=cte.transaction_id,
                last_transaction_hash=cte.hash,
                last_transaction_time=cte.block_time
              FROM cte
              WHERE
                id = cte.address_id
              `,
          )
          .queryContext(context.makeQueryContext());
      } else {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.address_id, transaction_id, hash, block_time
                FROM (
                  SELECT id1 AS address_id, MAX(id2) AS transaction_id
                  FROM address_to_transaction
                  WHERE
                    id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                    id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                  GROUP BY id1
                ) a
                JOIN 'transaction' b ON
                  a.transaction_id = b.id
              )
              UPDATE address SET
                last_transaction_id=(select cte.transaction_id from cte where address_id = address.id),
                last_transaction_hash=(select cte.hash from cte where address_id = address.id),
                last_transaction_time=(select cte.block_time from cte where address_id = address.id)
              WHERE
                id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
                `,
          )
          .queryContext(context.makeQueryContext());
      }
    }
  }
}
