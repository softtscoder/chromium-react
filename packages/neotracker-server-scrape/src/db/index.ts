export * from './BlockUpdater';
export * from './utils';
