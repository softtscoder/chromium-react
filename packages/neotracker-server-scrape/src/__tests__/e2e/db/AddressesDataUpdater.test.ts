import { Address as AddressModel, Transaction as TransactionModel } from '@neotracker/server-db';
import Knex from 'knex';
import { AddressesDataRevert, AddressesDataSave, AddressesDataUpdater } from '../../../db/AddressesDataUpdater';
import { AddressToTransactionUpdater } from '../../../db/AddressToTransactionUpdater';
import { Addresses } from '../../../types';
import { data, makeContext } from '../../data';
import { updaterUnitTest, UpdaterUtilTestOptions } from '../../data/updaterTestUtil';

const initialAddress: Partial<AddressModel> & Pick<AddressModel, 'id'> = {
  id: 'A'.repeat(34),
  transaction_hash: 'INIT'.repeat(16),
  transaction_count: '10',
  transfer_count: '5',
  block_id: 0,
  block_time: 15,
  aggregate_block_id: 0,
  transaction_id: '4',
};

const secondaryAddress: Partial<AddressModel> & Pick<AddressModel, 'id'> = {
  id: 'B'.repeat(34),
  transaction_hash: 'UNIT'.repeat(16),
  transaction_count: '10',
  transfer_count: '5',
  block_id: 0,
  block_time: 15,
  aggregate_block_id: 0,
  transaction_id: '9',
  last_transaction_hash: 'NINE'.repeat(16),
  last_transaction_id: '9',
  last_transaction_time: 15,
};

const addressToTransactions = {
  transactions: [
    {
      addressIDs: ['A'.repeat(34)],
      transactionID: '5',
    },
    {
      addressIDs: ['B'.repeat(34)],
      transactionID: '10',
    },
    {
      addressIDs: ['B'.repeat(34)],
      transactionID: '9',
    },
  ],
};

const transactions: ReadonlyArray<{ readonly id: string; readonly hash: string }> = [
  {
    id: '5',
    hash: 'XDXD'.repeat(16),
  },
  {
    id: '10',
    hash: 'TENT'.repeat(16),
  },
  {
    id: '9',
    hash: 'NINE'.repeat(16),
  },
];

const makeAddressesPatch = (addresses: ReadonlyArray<AddressModel>, transactionID: string, transactionHash: string) =>
  addresses.reduce<Addresses>(
    (acc: Addresses, addressModel) => ({
      ...acc,
      [addressModel.id.toString()]: {
        transactionCount: 5,
        transferCount: 5,
        transactionID,
        transactionHash,
      },
    }),
    {},
  );

const initiateTest = async (db: Knex) => {
  const context = makeContext({ db });

  const addressToTransactionUpdater = new AddressToTransactionUpdater();

  const references = await Promise.all([
    AddressModel.query(db)
      .context(context.makeQueryContext())
      .insertAndFetch(data.createAddress({ ...initialAddress })),
    AddressModel.query(db)
      .context(context.makeQueryContext())
      .insertAndFetch(data.createAddress({ ...secondaryAddress })),
  ]);

  await TransactionModel.insertAll(
    db,
    context.makeQueryContext(),
    transactions.map((transaction) => data.createTransaction({ ...transaction })),
  );

  await addressToTransactionUpdater.save(context, {
    ...addressToTransactions,
  });

  return { references };
};

const getInputs = (references: ReadonlyArray<AddressModel>): AddressesDataSave => ({
  addresses: makeAddressesPatch(references.slice(0, 1), '5', 'XDXD'.repeat(16)),
  blockIndex: 1,
  blockTime: 15,
});

const getSecondaryInputs = (references: ReadonlyArray<AddressModel>): AddressesDataSave => ({
  addresses: makeAddressesPatch(references.slice(1), '10', 'TENT'.repeat(16)),
  blockIndex: 1,
  blockTime: 15,
});

const getReversions = (references: ReadonlyArray<AddressModel>): AddressesDataRevert => ({
  addresses: makeAddressesPatch(references.slice(1), '10', 'TENT'.repeat(16)),
  transactionIDs: ['10'],
  blockIndex: 1,
});

const utilOptions: UpdaterUtilTestOptions<AddressesDataSave, AddressesDataRevert, ReadonlyArray<AddressModel>> = {
  name: `AddressesData Updater`,
  initiateTest,
  createUpdater: () => new AddressesDataUpdater(),
  getInputs,
  getSecondaryInputs,
  getReversions,
};

updaterUnitTest(utilOptions);
