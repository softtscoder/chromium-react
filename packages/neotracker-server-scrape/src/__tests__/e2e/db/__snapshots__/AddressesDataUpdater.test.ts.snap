// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressesData Updater AddressesData Updater: handles duplicate inserts/saves 1`] = `
Object {
  "action": Array [],
  "address": Array [
    Object {
      "aggregate_block_id": 1,
      "block_id": 0,
      "block_time": 15,
      "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "last_transaction_hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "last_transaction_id": "5",
      "last_transaction_time": 15,
      "transaction_count": "15",
      "transaction_hash": "INITINITINITINITINITINITINITINITINITINITINITINITINITINITINITINIT",
      "transaction_id": "4",
      "transfer_count": "10",
    },
    Object {
      "aggregate_block_id": 0,
      "block_id": 0,
      "block_time": 15,
      "id": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "last_transaction_hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "last_transaction_id": "9",
      "last_transaction_time": 15,
      "transaction_count": "10",
      "transaction_hash": "UNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNIT",
      "transaction_id": "9",
      "transfer_count": "5",
    },
  ],
  "addressToTransaction": Array [
    Object {
      "id1": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "id2": "5",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "10",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "9",
    },
  ],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "id": "5",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "id": "9",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "TENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENT",
      "id": "10",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
  ],
  "transactionInputOutput": Array [],
  "transfer": Array [],
}
`;

exports[`AddressesData Updater AddressesData Updater: handles insert/save 1`] = `
Object {
  "action": Array [],
  "address": Array [
    Object {
      "aggregate_block_id": 1,
      "block_id": 0,
      "block_time": 15,
      "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "last_transaction_hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "last_transaction_id": "5",
      "last_transaction_time": 15,
      "transaction_count": "15",
      "transaction_hash": "INITINITINITINITINITINITINITINITINITINITINITINITINITINITINITINIT",
      "transaction_id": "4",
      "transfer_count": "10",
    },
    Object {
      "aggregate_block_id": 0,
      "block_id": 0,
      "block_time": 15,
      "id": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "last_transaction_hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "last_transaction_id": "9",
      "last_transaction_time": 15,
      "transaction_count": "10",
      "transaction_hash": "UNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNIT",
      "transaction_id": "9",
      "transfer_count": "5",
    },
  ],
  "addressToTransaction": Array [
    Object {
      "id1": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "id2": "5",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "10",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "9",
    },
  ],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "id": "5",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "id": "9",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "TENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENT",
      "id": "10",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
  ],
  "transactionInputOutput": Array [],
  "transfer": Array [],
}
`;

exports[`AddressesData Updater AddressesData Updater: handles reverting inputs/saves 1`] = `
Object {
  "action": Array [],
  "address": Array [
    Object {
      "aggregate_block_id": 1,
      "block_id": 0,
      "block_time": 15,
      "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "last_transaction_hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "last_transaction_id": "5",
      "last_transaction_time": 15,
      "transaction_count": "15",
      "transaction_hash": "INITINITINITINITINITINITINITINITINITINITINITINITINITINITINITINIT",
      "transaction_id": "4",
      "transfer_count": "10",
    },
    Object {
      "aggregate_block_id": 0,
      "block_id": 0,
      "block_time": 15,
      "id": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "last_transaction_hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "last_transaction_id": "9",
      "last_transaction_time": 15,
      "transaction_count": "10",
      "transaction_hash": "UNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNIT",
      "transaction_id": "9",
      "transfer_count": "5",
    },
  ],
  "addressToTransaction": Array [
    Object {
      "id1": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "id2": "5",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "10",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "9",
    },
  ],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "id": "5",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "id": "9",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "TENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENT",
      "id": "10",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
  ],
  "transactionInputOutput": Array [],
  "transfer": Array [],
}
`;

exports[`AddressesData Updater AddressesData Updater: handles reverting inputs/saves 2`] = `
Object {
  "action": Array [],
  "address": Array [
    Object {
      "aggregate_block_id": 1,
      "block_id": 0,
      "block_time": 15,
      "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "last_transaction_hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "last_transaction_id": "5",
      "last_transaction_time": 15,
      "transaction_count": "15",
      "transaction_hash": "INITINITINITINITINITINITINITINITINITINITINITINITINITINITINITINIT",
      "transaction_id": "4",
      "transfer_count": "10",
    },
    Object {
      "aggregate_block_id": 1,
      "block_id": 0,
      "block_time": 15,
      "id": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "last_transaction_hash": "TENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENT",
      "last_transaction_id": "10",
      "last_transaction_time": 15,
      "transaction_count": "15",
      "transaction_hash": "UNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNIT",
      "transaction_id": "9",
      "transfer_count": "10",
    },
  ],
  "addressToTransaction": Array [
    Object {
      "id1": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "id2": "5",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "10",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "9",
    },
  ],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "id": "5",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "id": "9",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "TENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENT",
      "id": "10",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
  ],
  "transactionInputOutput": Array [],
  "transfer": Array [],
}
`;

exports[`AddressesData Updater AddressesData Updater: handles reverting inputs/saves 3`] = `
Object {
  "action": Array [],
  "address": Array [
    Object {
      "aggregate_block_id": 1,
      "block_id": 0,
      "block_time": 15,
      "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "last_transaction_hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "last_transaction_id": "5",
      "last_transaction_time": 15,
      "transaction_count": "15",
      "transaction_hash": "INITINITINITINITINITINITINITINITINITINITINITINITINITINITINITINIT",
      "transaction_id": "4",
      "transfer_count": "10",
    },
    Object {
      "aggregate_block_id": 0,
      "block_id": 0,
      "block_time": 15,
      "id": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "last_transaction_hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "last_transaction_id": "9",
      "last_transaction_time": 15,
      "transaction_count": "10",
      "transaction_hash": "UNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNITUNIT",
      "transaction_id": "9",
      "transfer_count": "5",
    },
  ],
  "addressToTransaction": Array [
    Object {
      "id1": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      "id2": "5",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "10",
    },
    Object {
      "id1": "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
      "id2": "9",
    },
  ],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "XDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXDXD",
      "id": "5",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "NINENINENINENINENINENINENINENINENINENINENINENINENINENINENINENINE",
      "id": "9",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
    Object {
      "attributes_raw": "0x1234",
      "block_id": 1,
      "block_time": 15,
      "gas": null,
      "hash": "TENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENTTENT",
      "id": "10",
      "index": 1,
      "network_fee": "0.00000000",
      "nonce": null,
      "pubkey": null,
      "result_raw": null,
      "script": null,
      "scripts_raw": "0xab38352559b8b203bde5fddfa0b07d8b25111111",
      "size": 10,
      "system_fee": "0.00000000",
      "type": "ContractTransaction",
      "version": 0,
    },
  ],
  "transactionInputOutput": Array [],
  "transfer": Array [],
}
`;
